src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -296,13 +315,20 @@ private async ValueTask<FlushResult> ProcessDataWrites()
 
                     if (readResult.IsCompleted && _stream.Trailers?.Count > 0)
                     {
+                        // Output is ending and there are trailers to write
+                        // Write any remaining content then write trailers
                         if (readResult.Buffer.Length > 0)
                         {
                             flushResult = await _frameWriter.WriteDataAsync(_streamId, _flowControl, readResult.Buffer, endStream: false);
                         }
 
                         flushResult = await _frameWriter.WriteResponseTrailers(_streamId, _stream.Trailers);
                     }
+                    else if (readResult.IsCompleted && _streamEnded)
+                    {
+                        // Headers have already been written and there is no other content to write
+                        flushResult = await _frameWriter.FlushAsync(outputAborter: null, cancellationToken: default);
```

Comment:
Where else is FlushAsync called? Can we add a method to FrameWriter that takes no parameters instead?

Reply:
Everything calls down to TimingPipeFlusher.FlushAsync with these arguments in the end. I think it could change to `Http2OutputProducer.FlushAsync` but that will eventually call back to this method anyway 🤷‍♂